cli.go

```diff
@@ -268,10 +342,109 @@ func (c *CLI) init() {
 		}
 	}
 
+	// Setup autocomplete if we have it enabled. We have to do this after
+	// the command tree is setup so we can use the radix tree to easily find
+	// all subcommands.
+	if c.Autocomplete {
+		c.initAutocomplete()
+	}
+
 	// Process the args
 	c.processArgs()
 }
 
+func (c *CLI) initAutocomplete() {
+	if c.AutocompleteInstall == "" {
+		c.AutocompleteInstall = defaultAutocompleteInstall
+	}
+
+	if c.AutocompleteUninstall == "" {
+		c.AutocompleteUninstall = defaultAutocompleteUninstall
+	}
+
+	if c.autocompleteInstaller == nil {
+		c.autocompleteInstaller = &realAutocompleteInstaller{}
+	}
+
+	// Build the root command
+	cmd := c.initAutocompleteSub("")
+
+	// For the root, we add the global flags
+	cmd.GlobalFlags = map[string]complete.Predictor{
+		"-" + c.AutocompleteInstall:   complete.PredictNothing,
+		"-" + c.AutocompleteUninstall: complete.PredictNothing,
+		"-help":    complete.PredictNothing,
+		"-version": complete.PredictNothing,
+	}
+	for k, v := range c.AutocompleteGlobalFlags {
+		cmd.GlobalFlags[k] = v
+	}
+
+	c.autocomplete = complete.New(c.Name, cmd)
+}
+
+func (c *CLI) initAutocompleteSub(prefix string) complete.Command {
```

Comment:
Probably worth a comment that this is a recursive call to walk all the subcommands